shell/platform/fuchsia/dart_runner/dart_component_controller_v2.cc

```diff
@@ -0,0 +1,489 @@
+    return false;
+  }
+
+  const zx_status_t ns_create_status = fdio_ns_create(&namespace_);
+  if (ns_create_status != ZX_OK) {
+    FX_LOGF(ERROR, LOG_TAG, "Failed to create namespace: %s",
+            zx_status_get_string(ns_create_status));
+  }
+
+  dart_utils::RunnerTemp::SetupComponent(namespace_);
+
+  // Bind each directory in start_info's namespace to the controller's namespace
+  // instance.
+  for (auto& ns_entry : *start_info_.mutable_ns()) {
+    // TODO(akbiggs): Under what circumstances does a namespace entry not have a
+    // path or directory? Should we log an error for these?
+    if (!ns_entry.has_path() || !ns_entry.has_directory()) {
+      continue;
+    }
+
+    if (ns_entry.path() == kTmpPath) {
+      // /tmp is covered by the local memfs.
+      continue;
+    }
+
+    // TODO(akbiggs): Why is it okay/desirable to transfer ownership of these?
+    fidl::InterfaceHandle<::fuchsia::io::Directory> dir =
+        std::move(*ns_entry.mutable_directory());
+    const std::string path = std::move(*ns_entry.mutable_path());
+
+    const zx_status_t ns_bind_status =
+        fdio_ns_bind(namespace_, path.c_str(), dir.TakeChannel().release());
+    if (ns_bind_status != ZX_OK) {
+      FX_LOGF(ERROR, LOG_TAG, "Failed to bind %s to namespace: %s",
+              path.c_str(), zx_status_get_string(ns_bind_status));
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool DartComponentControllerV2::SetUpFromKernel() {
+  dart_utils::MappedResource manifest;
+  if (!dart_utils::MappedResource::LoadFromNamespace(
+          namespace_, data_path_ + "/app.dilplist", manifest)) {
+    return false;
+  }
+
+  if (!dart_utils::MappedResource::LoadFromNamespace(
+          nullptr, "/pkg/data/isolate_core_snapshot_data.bin",
+          isolate_snapshot_data_)) {
+    return false;
+  }
+  if (!dart_utils::MappedResource::LoadFromNamespace(
+          nullptr, "/pkg/data/isolate_core_snapshot_instructions.bin",
+          isolate_snapshot_instructions_, true /* executable */)) {
+    return false;
+  }
+
+  if (!CreateIsolate(isolate_snapshot_data_.address(),
+                     isolate_snapshot_instructions_.address())) {
+    return false;
+  }
+
+  Dart_EnterScope();
+
+  std::string str(reinterpret_cast<const char*>(manifest.address()),
+                  manifest.size());
+  Dart_Handle library = Dart_Null();
+  for (size_t start = 0; start < manifest.size();) {
+    size_t end = str.find("\n", start);
+    if (end == std::string::npos) {
+      FX_LOG(ERROR, LOG_TAG, "Malformed manifest");
+      Dart_ExitScope();
+      return false;
+    }
+
+    std::string path = data_path_ + "/" + str.substr(start, end - start);
+    start = end + 1;
+
+    dart_utils::MappedResource kernel;
+    if (!dart_utils::MappedResource::LoadFromNamespace(namespace_, path,
+                                                       kernel)) {
+      FX_LOGF(ERROR, LOG_TAG, "Failed to find kernel: %s", path.c_str());
```

Comment:
nit: I think this might lead to confusion in v2. In CFv2 we have locked down which processes can mark a vmo executable. There is a security allowlist that will let us do this but that will only be enabled for certain builds and is something we have to figure out longer term. I think this message should be updated to be something more like cannot load kernel from namespace, this could be due to a missing kernel or invalid privileges.

Reply:
Done.